### PR TITLE
YouTube preview on Kifi.com

### PR DESCRIPTION
**Pull request by andrewconner**
YouTube screenshots are fairly borked because our provider is a) in Germany, so many videos are blocked, b) does not play flash videos, so shows a black box.

This change shows a custom youtube player instead of the screen shot. Danny got very excited about the possibility of this a few days ago, so it's an easy win if you guys are happy with it.

image

### CODE REVIEW

**Comment by andrewconner** [comment]
@2is10 You're probably the one to green light it.

File: kifi-backend/modules/shoebox/public/js/main.js [hunk 1]
```diff
@@ -376,11 +376,26 @@ $(function () {
...
-					$pic.find('.page-pic-soon').addClass('showing');
-				});
+				var skipImage = false;
+
+				if (o.url.indexOf('://www.youtube.com/') > -1) {
+					var youtubeRegex = /https?:\/\/(?:[0-9A-Z-]+\.)?(?:youtu\.be\/|youtube\.com\S*[^\w\-\s])([\w\-]{11})(?=[^\w\-]|$)[?=&+%\w.-]*/i;
+					var match = url.match(youtubeRegex);
+					if (match && match.length == 2) {
+						var vID = match[1];
+						var embedHtml = '<object width= "434" height="226"><param name="movie" value="http://www.youtube.com/v/' + vID + '&rel=0&theme=light&showinfo=0&disablekb=1&modestbranding=1&controls=0&hd=1&autohide=1&color=white"></param><param name="allowFullScreen" value="true"></param><param name="allowscriptaccess" value="always"></param><embed src="http://www.youtube.com/v/' + vID + '&rel=0&theme=light&showinfo=0&disablekb=1&modestbranding=1&controls=0&hd=1&autohide=1&color=white" type="application/x-shockwave-flash" allowfullscreen="true" width="434" height="226" allowscriptaccess="always"></embed></object>';
```
Comment (2is10) [inline kifi-backend/modules/shoebox/public/js/main.js]: The size of the preview image in the DOM is responsive (grows and shrinks with window width). Not sure it’s a good idea to assume any fixed size for the video. Does a YouTube video `<object>` handle resizing gracefully during play?

**Comment by 2is10** [comment]
Great feature if we can figure out the size issue. I’d like to review it more carefully if you work that out.

File: kifi-backend/modules/shoebox/public/js/main.js (hunk 1 above)
Comment (andrewconner) [inline kifi-backend/modules/shoebox/public/js/main.js]: You're right, I missed this. Just pushed a fix. I feel it's messy, but it works. Can you think of a cleaner way?

**Comment by eishay** [comment]
cool!

File: kifi-backend/modules/shoebox/public/css/main.css [hunk 2]
```diff
@@ -3722,6 +3722,33 @@ a.page-url[href]:hover {
...
+
+.page-pic-special .youtube {
+  position: relative;
+  padding-bottom: 56.25%; /* 16:9 */
+  height: 0;
+  z-index: 2;
+}
+
+.page-pic-special .youtube object,
+.page-pic-special .youtube embed,  {
```
Comment (2is10) [inline kifi-backend/modules/shoebox/public/css/main.css]: remove trailing comma

File: kifi-backend/modules/shoebox/public/js/main.js [hunk 3]
```diff
@@ -376,11 +376,27 @@ $(function () {
...
-				$.postJson(xhrBase + '/keeps/screenshot', {url: o.url}, function (data) {
-					$pic.css('background-image', 'url(' + data.url + ')');
-				}).fail(function () {
-					$pic.find('.page-pic-soon').addClass('showing');
-				});
+				var skipImage = false;
+
+				if (o.url.indexOf('://www.youtube.com/') > -1) {
+					var youtubeRegex = /https?:\/\/(?:[0-9A-Z-]+\.)?(?:youtu\.be\/|youtube\.com\S*[^\w\-\s])([\w\-]{11})(?=[^\w\-]|$)[?=&+%\w.-]*/i;
+					var match = url.match(youtubeRegex);
```
Comment (2is10) [inline kifi-backend/modules/shoebox/public/js/main.js:383]: odd to use `url` here and `o.url` two lines earlier. probably better to use the local var `url` consistently.

File: kifi-backend/modules/shoebox/public/js/main.js [hunk 4]
```diff
@@ -376,11 +376,27 @@ $(function () {
...
 				var $pic = $('.page-pic'), $chatter = $('.page-chatter-messages');
-				$.postJson(xhrBase + '/keeps/screenshot', {url: o.url}, function (data) {
-					$pic.css('background-image', 'url(' + data.url + ')');
-				}).fail(function () {
-					$pic.find('.page-pic-soon').addClass('showing');
-				});
+				var skipImage = false;
+
+				if (o.url.indexOf('://www.youtube.com/') > -1) {
+					var youtubeRegex = /https?:\/\/(?:[0-9A-Z-]+\.)?(?:youtu\.be\/|youtube\.com\S*[^\w\-\s])([\w\-]{11})(?=[^\w\-]|$)[?=&+%\w.-]*/i;
```
Comment (2is10) [inline kifi-backend/modules/shoebox/public/js/main.js:382]: the regex seems to look for youtu.be URLs too, but those won’t get past the check on the previous line

File: kifi-backend/modules/shoebox/public/js/main.js [hunk 5]
```diff
@@ -376,11 +376,27 @@ $(function () {
...
+					var match = url.match(youtubeRegex);
+					if (match && match.length == 2) {
+						var vID = match[1];
+						var embedHtml = '<div class="youtube"><object><param name="movie" value="http://www.youtube.com/v/' + vID + '&rel=0&theme=light&showinfo=0&disablekb=1&modestbranding=1&controls=0&hd=1&autohide=1&color=white"></param><param name="allowFullScreen" value="true"></param><param name="allowscriptaccess" value="always"></param><embed src="http://www.youtube.com/v/' + vID + '&rel=0&theme=light&showinfo=0&disablekb=1&modestbranding=1&controls=0&hd=1&autohide=1&color=white" type="application/x-shockwave-flash" allowfullscreen="true" style="width:100%; height: 100%;" allowscriptaccess="always"></embed></object></div>';
```
Comment (2is10) [inline kifi-backend/modules/shoebox/public/js/main.js]: Let’s use either `object` or `embed` (not both). All browsers we care about (and their grandparents) support either.

File: kifi-backend/modules/shoebox/public/js/main.js [hunk 6]
```diff
@@ -376,11 +376,27 @@ $(function () {
...
+						$('.page-pic-special').html(embedHtml).addClass('page-pic-special-cell').show();
```
Comment (2is10) [inline kifi-backend/modules/shoebox/public/js/main.js:387]: Here it seems you’re adding the `page-pic-special-cell` class to the element with class `page-pic-special`, but in the stylesheet I see the descendant selector `.page-pic-special .page-pic-special-cell` so I’m a little confused. Remove the space in the selector if you’re trying to match elements with both classes.

**Comment by 2is10** [comment]
I may try to simplify or adjust this a bit after it’s committed, but if you’re confident it works, go for it!

Probably a good idea to try it in at least two browsers. Chrome/Firefox/Safari is a good habit, although I know there are currently some things wrong with our site in Firefox & Safari. Still a good idea not to introduce new cross-browser compatibility issues. Thanks.

**Comment by andrewconner** [comment]
Thanks for the review. Tested in Chrome, Firefox, and Safari and it works. Merging — hope this doesn't give you a headache cleaning it up :P

**Pull request merged**